### PR TITLE
VA-019:Save the audio in MongoDB

### PR DESCRIPTION
**Pull request by J-Ivan-MC**

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -22,23 +22,25 @@ services:
...
+    image: mongo
+    restart: always
+    volumes:
+      - mongo_data:/data/db
     ports:
-      - 8888:8888
+      - 27017:27017
+    environment:
+      MONGO_INITDB_ROOT_USERNAME: VirtualAssistant60089
+      MONGO_INITDB_ROOT_PASSWORD: rGYwVKko8ihLmyaC
```
Comment (zamoranoalicia) [inline docker-compose.yml:43]: same here

File: src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import org.springframework.stereotype.Repository;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.util.Base64;
+import java.util.List;
+
+@Repository
+public class RecordingRepositoryImpl implements RecordingRepo {
+
+    @Autowired
```
Comment (zamoranoalicia) [inline src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java:19]: please dont use @Autowire here

Comment (zamoranoalicia) [reply src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java:19]: Spring injects the dependencies automatically

File: src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    MongoTemplate mongoTemplate;
+
+    @Override
+    public Recording getRecording(String idRecording) {
+        Query query = new Query(Criteria.where("idRecording").is(idRecording));
```
Comment (zamoranoalicia) [inline src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java:24]: You are repiting this same portion of code, in almost all of the methods, so please abstract this Query creating in a separated method

File: src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        Recording document = mongoTemplate.findOne(query, Recording.class);
+        return document;
+    }
+
+    @Override
+    public boolean deleteRecording(String idRecording) {
+        Query query = new Query(Criteria.where("idRecording").is(idRecording));
```
Comment (zamoranoalicia) [inline src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java:31]: same here, please apply DRY principle https://www.baeldung.com/cs/dry-software-design-principle

File: src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+        Recording recordingToDelete = mongoTemplate.findOne(query, Recording.class);
+        if (recordingToDelete != null) {
```
Comment (zamoranoalicia) [inline src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java:33]: invert this comparison, or use ofNullable(recordingToDelete).ifPresent()

File: src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        Document metadata = generateDocumentRecording(audioFile);
+        Recording recording = new Recording(idUser, idChat, metadata);
+        return mongoTemplate.save(recording);
+    }
+
+    private Document generateDocumentRecording (MultipartFile file){
+        try {
+            byte[] audioBytes = file.getBytes();
+            String encodedAudio = Base64.getEncoder().encodeToString(audioBytes);
+            Document audioDocument = new Document("audio", encodedAudio);
```
Comment (zamoranoalicia) [inline src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java:62]: use constant here for "audio"

File: src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            return audioDocument;
+        } catch (Exception e) {
+            throw new RuntimeException(e + " Exception: Error in generate Document");
```
Comment (zamoranoalicia) [inline src/main/java/org/fundacionjala/virtualassistant/repository/mongo/RecordingRepositoryImpl.java:65]: use especific exception here

File: src/main/java/org/fundacionjala/virtualassistant/services/mongo/RecordingService.java [hunk 8]
```diff
@@ -0,0 +1,21 @@
+package org.fundacionjala.virtualassistant.services.mongo;
+
+import org.fundacionjala.virtualassistant.models.mongo.Recording;
+import org.springframework.web.multipart.MultipartFile;
+
+import java.util.List;
+
+
```
Comment (zamoranoalicia) [inline src/main/java/org/fundacionjala/virtualassistant/services/mongo/RecordingService.java:8]: In this case is not necessary to create an interface for the service

File: src/main/resources/application.properties [hunk 9]
```diff
@@ -6,3 +6,16 @@ spring.jpa.hibernate.ddl-auto=update
 spring.jpa.open-in-view=false
 server.port=8081
 spring.flyway.enabled=false
+
+# MongoDB settings
+spring.data.mongodb.username=VirtualAssistant60089
```
Comment (zamoranoalicia) [inline src/main/resources/application.properties:11]: dont expose users and passwords

File: src/main/resources/application.properties [hunk 10]
```diff
@@ -6,3 +6,16 @@ spring.jpa.hibernate.ddl-auto=update
...
+spring.data.mongodb.password=rGYwVKko8ihLmyaC
```
Comment (zamoranoalicia) [inline src/main/resources/application.properties:12]: set this values as environment variables

File: src/test/java/org/fundacionjala/virtualassistant/services/mongo/RecordingServiceImplTest.java [hunk 11]
```diff
@@ -0,0 +1,40 @@
...
+import org.fundacionjala.virtualassistant.models.mongo.Recording;
+import org.fundacionjala.virtualassistant.repository.mongo.RecordingRepo;
+import org.fundacionjala.virtualassistant.repository.mongo.RecordingRepositoryImpl;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+import org.springframework.mock.web.MockMultipartFile;
+import org.springframework.web.multipart.MultipartFile;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.mockito.Mockito.*;
```
Comment (zamoranoalicia) [inline src/test/java/org/fundacionjala/virtualassistant/services/mongo/RecordingServiceImplTest.java:13]: dont use wildcards when importing classes use specific class names

**Review by zamoranoalicia — CHANGES_REQUESTED** [review]

**Comment by santihsdev** [comment]
[VA-019 [SP-5]: Save the audio in MongoDB](https://trello.com/c/WkH55tcU/35-va-019-sp-5-save-the-audio-in-mongodb)

**Pull request merged**